pkg/identity/github/principal_test.go

```diff
@@ -0,0 +1,280 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package github
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"encoding/asn1"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"testing"
+	"unsafe"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/sigstore/fulcio/pkg/identity"
+)
+
+func TestWorkflowPrincipalFromIDToken(t *testing.T) {
+	tests := map[string]struct {
+		Claims          map[string]interface{}
+		ExpectPrincipal workflowPrincipal
+		WantErr         bool
+	}{
+		`Valid token authenticates with correct claims`: {
+			Claims: map[string]interface{}{
+				"aud":              "sigstore",
+				"event_name":       "push",
+				"exp":              0,
+				"iss":              "https://token.actions.githubusercontent.com",
+				"job_workflow_ref": "sigstore/fulcio/.github/workflows/foo.yaml@refs/heads/main",
+				"ref":              "refs/heads/main",
+				"repository":       "sigstore/fulcio",
+				"sha":              "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+				"sub":              "repo:sigstore/fulcio:ref:refs/heads/main",
+				"workflow":         "foo",
+			},
+			ExpectPrincipal: workflowPrincipal{
+				issuer:     "https://token.actions.githubusercontent.com",
+				subject:    "repo:sigstore/fulcio:ref:refs/heads/main",
+				url:        "https://github.com/sigstore/fulcio/.github/workflows/foo.yaml@refs/heads/main",
+				sha:        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+				trigger:    "push",
+				repository: "sigstore/fulcio",
+				workflow:   "foo",
+				ref:        "refs/heads/main",
+			},
+			WantErr: false,
+		},
+		`Token missing job_workflow_ref claim should be rejected`: {
+			Claims: map[string]interface{}{
+				"aud":        "sigstore",
+				"event_name": "push",
+				"exp":        0,
+				"iss":        "https://token.actions.githubusercontent.com",
+				"ref":        "refs/heads/main",
+				"repository": "sigstore/fulcio",
+				"sha":        "bf96275471e83ff04ce5c8eb515c04a75d43f854",
+				"sub":        "repo:sigstore/fulcio:ref:refs/heads/main",
+				"workflow":   "foo",
+			},
+			WantErr: true,
```

Comment:
What do you think about also checking the contents of the error message to check that the missing field is mentioned in the error message? This would be helpful to make sure that the test is constructed with the correct missing field.